executor/internal/querywatch/query_watch.go

```diff
@@ -16,13 +16,143 @@ package querywatch
+			}
+			record.ResourceGroupName = name
+		} else {
+			record.ResourceGroupName = op.StrValue.L
+		}
+	case ast.QueryWatchAction:
+		record.Action = rmpb.RunawayAction(op.IntValue)
+	case ast.QueryWatchType:
+		expr, err := expression.RewriteAstExpr(sctx, op.ExprValue, nil, nil, false)
+		if err != nil {
+			return err
+		}
+		strval, isNull, err := expr.EvalString(sctx, chunk.Row{})
+		if err != nil {
+			return err
+		}
+		if isNull {
+			return errors.Errorf("invalid watch text expression")
+		}
+		record.Watch = rmpb.RunawayWatchType(op.IntValue)
+		if op.BoolValue {
+			p := parser.New()
+			stmts, _, err := p.ParseSQL(strval)
+			if err != nil {
+				return err
+			}
+			if len(stmts) != 1 {
+				return errors.Errorf("only support one SQL")
+			}
+			sql := stmts[0].Text()
+			switch model.RunawayWatchType(op.IntValue) {
+			case model.WatchNone:
+				return errors.Errorf("watch type must be specified")
+			case model.WatchExact:
+				record.WatchText = sql
+			case model.WatchSimilar:
+				_, digest := parser.NormalizeDigest(sql)
+				record.WatchText = digest.String()
+			case model.WatchPlan:
+				sqlExecutor := newSctx.(sqlexec.SQLExecutor)
+				if _, err := sqlExecutor.ExecuteInternal(ctx, fmt.Sprintf("explain %s", stmts[0].Text())); err != nil {
+					return err
+				}
+				_, digest := newSctx.GetSessionVars().StmtCtx.GetPlanDigest()
+				if digest == nil {
+					return errors.Errorf("no plan digest")
+				}
+				record.WatchText = digest.String()
+			}
+		} else {
+			if len(strval) != 64 {
+				return errors.Errorf("digest format error")
+			}
+			record.WatchText = strval
+		}
+	}
+	return nil
+}
+
+// fromQueryWatchOptionList is used to create a QuarantineRecord with some QueryWatchOption ast nodes.
+func fromQueryWatchOptionList(ctx context.Context, sctx, newSctx sessionctx.Context, optionList []*ast.QueryWatchOption) (*resourcegroup.QuarantineRecord, error) {
+	record := &resourcegroup.QuarantineRecord{
+		Source:    resourcegroup.ManualSource,
+		StartTime: time.Now(),
+		EndTime:   resourcegroup.NullTime,
+	}
+	for _, op := range optionList {
+		if err := setWatchOption(ctx, sctx, newSctx, record, op); err != nil {
+			return nil, err
+		}
+	}
+	return record, nil
+}
+
+// validateWatchRecord follows several designs:
+//  1. If no resource group is set, the default resource group is used
+//  2. If no action is specified, the action of the resource group is used. If no, an error message is displayed.
+func validateWatchRecord(record *resourcegroup.QuarantineRecord, client *rmclient.ResourceGroupsController) error {
+	if len(record.ResourceGroupName) == 0 {
+		record.ResourceGroupName = resourcegroup.DefaultResourceGroupName
+	}
+	rg, err := client.GetResourceGroup(record.ResourceGroupName)
+	if err != nil {
+		return err
+	}
+	if rg == nil {
+		return infoschema.ErrResourceGroupNotExists.GenWithStackByArgs(record.ResourceGroupName)
+	}
+	if record.Action == rmpb.RunawayAction_NoneAction {
+		if rg.RunawaySettings == nil {
```

Comment:
What happens if when the RunawaySettings is not nil when adding the record and then changed to none, then what is the action of the record

Reply:
This watch will be uselss but not be removed from watch list. You can see `BeforeExecutor`, If no match action, it will do nothing.